### PR TITLE
feat(shop-online): made the shop and owned item persistant in Firebase

### LINKED ISSUE #194: Bug on the shop
**Description:** 

When you buy something, you can buy it again. But when you look at the items you have, you only find one instance while you bought two of the same item. To fix this, simply implement persistent ownership for the items, and remove the possibility to buy twice the same one.

**Criteria:**

- [x] Buying an item gives you a persistent ownership
- [x] It's not possible to buy the same item twice

### LINKED ISSUE #239: Make shop and owned item persistant in firebase

### PR DESCRIPTION
**Pull request by khlilinfo**
# Shop Repository with Firebase Persistence

### What problem are we solving?

The app lacked a persistent, remote shop system for cosmetic items. Items were only stored locally in code, making it impossible to:
- Dynamically add/update shop items without app updates
- Share shop inventory across all users
- Store user purchases persistently
- Load shop data remotely

### Brief background

Previously, shop items were hardcoded in the ViewModel with no persistence layer. User purchases were stored only in the profile's `accessories` list as strings, with no proper tracking or validation. This made it difficult to manage the shop inventory and track purchases reliably.

---

### Core Implementation

**Repository Pattern:**
- ✅ `ShopRepository` interface defining the shop data contract
- ✅ `FakeShopRepository` for in-memory testing (no Firebase)
- ✅ `FirestoreShopRepository` for Firebase-backed persistence
- ✅ Integrated with `RepositoriesProvider` (Fake & Firestore)

**Data Model:**
- ✅ `CosmeticItem` data class with support for:
  - Local drawable resources (fallback)
  - Remote image URLs (Firebase Storage ready)
  - Owned status tracking
  - Price, name, and ID fields

**ViewModel Integration:**
- ✅ `ShopViewModel` refactored to use `ShopRepository`
- ✅ Automatic seeding on first use
- ✅ State flow for reactive UI updates
- ✅ Purchase logic with coin deduction and Firebase persistence

**Admin Utilities:**
- ✅ `SeedShopItems` utility for manual shop initialization
- ✅ Methods: `seedIfNeeded()`, `forceReseed()`, `checkIfItemsExist()`, `getItemCount()`, `listAllItems()`

### Firestore Schema

**Collection: `/shopItems/{itemId}`** (Shared across all users)
```javascript
{
  id: "glasses",
  name: "Cool Shades",
  price: 200,
  imageResName: "shop_cosmetic_glasses",
  imageUrl: "https://...", // Optional remote URL
  description: "Stylish sunglasses for your EduMon",
  rarity: "common",
  createdAt: Timestamp
}
```

**Collection: `/users/{uid}/ownedCosmetics/{itemId}`** (Per-user purchases)
```javascript
{
  itemId: "glasses",
  purchasedAt: Timestamp
}
```

### Firebase Rules Updates

Added security rules for shop collections:
```javascript
// Shop items (read by all, write restricted)
match /shopItems/{itemId} {
  allow read: if signedIn();
  allow create, update: if signedIn(); // TODO: Restrict to admin
  allow delete: if false;
}

// User purchases (private per user)
match /users/{uid}/ownedCosmetics/{itemId} {
  allow read: if isOwner(uid);
  allow create: if isOwner(uid) && request.resource.data.itemId == itemId;
  allow update, delete: if false; // Immutable purchases
}
```

**Security Features:**
- ✅ User isolation (users can't see each other's purchases)
- ✅ Immutable purchases (no refunds)
- ✅ Data integrity validation (`itemId` must match document ID)

---

### Key Decisions & Trade-offs

1. **Auto-seeding Strategy:**
   - Repository automatically seeds default items if `shopItems` collection is empty
   - Seeds on first `getItems()` call (typically when Shop screen opens)
   - Trade-off: Requires Firebase write permissions for users (can be restricted to admin later)

2. **Hybrid Image Support:**
   - Items store both `imageResName` (local drawable) and `imageUrl` (remote)
   - Falls back to local drawable if remote URL unavailable
   - Allows gradual migration to Firebase Storage without breaking existing functionality

3. **Repository Abstraction:**
   - Clean interface allows swapping between Fake and Firestore implementations
   - Tests use `FakeShopRepository` for fast unit tests
   - Emulator tests use `FirestoreShopRepository` with real Firebase

4. **Purchase Flow:**
   - Coins deducted from `ProfileRepository` (profile data)
   - Purchase recorded in `ShopRepository` (ownership tracking)
   - Two separate concerns, two repositories (Single Responsibility Principle)

### Concurrency/Data Consistency

- **Thread-safe:** All Firebase operations use `withContext(dispatchers.io)`
- **Atomic writes:** Batch operations used for seeding (all-or-nothing)
- **Optimistic updates:** Local state updated immediately, then synced with Firebase
- **Flow synchronization:** `_items` StateFlow updated after Firebase operations complete
- **No transactions needed:** User purchases are independent (no inventory depletion)

### Performance Considerations

- **Caching:** Items loaded once and cached in `_items` StateFlow
- **Lazy loading:** Repository doesn't fetch until `getItems()` is called
- **Batch operations:** All 6 items seeded in single batch write
- **Minimal reads:** Owned status fetched once per session via `refreshOwnedStatus()`
- **No real-time listeners:** Reduces Firebase quota usage (consider adding later for dynamic prices)

---

### Unit Tests (3 files, 45 tests)

**`FakeShopRepositoryTest.kt`** (24 tests)
- All repository operations (get, purchase, owned tracking)
- State management and flow updates
- Edge cases (duplicate purchases, all items owned)

**`FirestoreShopRepositoryTest.kt`** (8 tests)
- Static method validation (`defaultCosmetics()`)
- Price correctness, item structure
- Unique IDs and names

**`ShopViewModelTest.kt`** (16 tests, updated)
- Coin flow from profile
- Items flow from shop repository
- Purchase success/failure scenarios
- Multiple purchases handling
- Init refreshes owned status

### Instrumented Tests (1 file, 25 tests)

**`FirestoreShopRepositoryEmulatorTest.kt`** (25 tests)
- Real Firebase Emulator integration
- Auto-seeding verification
- Purchase persistence across repository instances
- User isolation (different users have separate purchases)
- Multi-user scenarios
- Refresh synchronization

### UI Tests (1 file, 7 tests, updated)

**`ShopScreenTest.kt`** (7 tests)
- Shop UI rendering with correct items
- Purchase button interactions
- Owned item badge display
- Drawable resource references fixed

**Coverage Summary:**
- ✅ Repository logic: 100% (all operations tested)
- ✅ Firebase integration: End-to-end with emulator
- ✅ ViewModel: All purchase flows covered
- ✅ UI: Basic rendering verified
- ✅ Edge cases: User isolation, concurrent ops, all items owned

---

## Screenshots / Demos

**Shop Screen:**
- Displays 6 cosmetic items with prices
- Shows owned badge for purchased items
- "Cool Shades", "Wizard Hat", "Red Scarf", "Cyber Wings", "Epic Aura" (1500 coins), "Hero Cape"

**Firebase Console:**
image

image

---

## Checklist

- ✅ Code compiles without errors
- ✅ Repository interface defined with all CRUD operations
- ✅ Fake and Firestore implementations complete
- ✅ Integrated with provider pattern
- ✅ ViewModel refactored to use repository
- ✅ Auto-seeding logic implemented
- ✅ Firebase rules updated and secure
- ✅ Unit tests: 45 tests covering all scenarios
- ✅ Emulator tests: 25 tests with real Firebase
- ✅ UI tests updated with correct drawable references
- ✅ Manual testing completed (auto-seed, purchase, persistence)
- ✅ User isolation verified (multi-user tests pass)
- ✅ Documentation: `SeedShopItems.kt` utility documented
- ⏳ Production deployment: Review Firebase rules (restrict shop item writes to admin)
- ⏳ Firebase Storage: Upload images for remote URLs (future enhancement)
- ⏳ Admin panel: Create UI for managing shop items (future enhancement)

---

### Migration from Old System

**Breaking Changes:**
- Old system stored purchases as `accessories: List<String>` with format `"owned:itemId"`
- New system uses dedicated `ownedCosmetics` subcollection
- Migration script needed if existing users have purchased items

**Migration Strategy (if needed):**
```kotlin
suspend fun migrateOldPurchases(userProfile: UserProfile, userId: String) {
    val db = FirebaseFirestore.getInstance()
    val ownedCol = db.collection("users").document(userId).collection("ownedCosmetics")
    
    userProfile.accessories
        .filter { it.startsWith("owned:") }
        .map { it.removePrefix("owned:") }
        .forEach { itemId ->
            ownedCol.document(itemId).set(mapOf(
                "itemId" to itemId,
                "purchasedAt" to FieldValue.serverTimestamp(),
                "migratedFrom" to "profile.accessories"
            ))
        }
}
```

### Known Limitations

1. **No inventory limits:** Items can be purchased by unlimited users
2. **No price history:** Price changes don't track old purchases
3. **Single currency:** Only supports coins (no gems, premium currency)
4. **Static catalog:** No seasonal/limited items without code update
5. **No bundles:** Can't purchase multiple items at once with discount

### Future Enhancements

- [ ] Real-time listeners for dynamic pricing
- [ ] Firebase Storage integration for remote images
- [ ] Admin dashboard for shop management
- [ ] Purchase history/analytics
- [ ] Limited-time items with expiry
- [ ] Item categories and filtering
- [ ] Wish list functionality
- [ ] Gift items to friends
- [ ] Referral rewards (free items)

---

## Issue Reference

Closes #194 (Shop System with Firebase Persistence)
Closes #239 

### Manual Testing:

1. **First-time user flow:**
   ```
   - Launch app
   - Navigate to Shop screen
   - Verify 6 items appear
   - Check Firebase Console → `shopItems` collection exists
   ```

2. **Purchase flow:**
   ```
   - Click item → "Start" button
   - Verify coins deducted from profile
   - Verify "Owned" badge appears
   - Restart app → item still owned
   ```

3. **Multi-user isolation:**
   ```
   - User A purchases "glasses"
   - Sign out, sign in as User B
   - User B should NOT see "glasses" as owned
   ```

### CODE REVIEW

File: app/src/androidTest/java/com/android/sample/ui/shop/repository/FirestoreShopRepositoryEmulatorTest.kt
Comment (Adessrrrr) [inline app/src/androidTest/java/com/android/sample/ui/shop/repository/FirestoreShopRepositoryEmulatorTest.kt:1]: It's very good that you added a lot of tests like this, it leads to having 98.4% of line coverage which is incredible. Nothing to say for this one, tests are good.

File: app/src/androidTest/java/com/android/sample/ui/shop/ShopScreenTest.kt
Comment (Adessrrrr) [inline app/src/androidTest/java/com/android/sample/ui/shop/ShopScreenTest.kt:1]: Nothing to say, good.

File: app/src/main/java/com/android/sample/ui/shop/repository/FakeShopRepository.kt [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+  private fun updateItemsOwnedStatus() {
+    _items.update { list -> list.map { item -> item.copy(owned = ownedIds.contains(item.id)) } }
+  }
+
+  companion object {
+    /** Default cosmetic items available in the shop. */
+    fun defaultCosmetics(): List<CosmeticItem> =
+        listOf(
+            CosmeticItem("glasses", "Cool Shades", 200, R.drawable.shop_cosmetic_glasses),
```
Comment (Adessrrrr) [inline app/src/main/java/com/android/sample/ui/shop/repository/FakeShopRepository.kt]: Please try to avoid hardcoded strings and magic numbers, same for lines 45,46,47,48,49.

File: app/src/main/java/com/android/sample/ui/shop/repository/FirestoreShopRepository.kt [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        val defaults = defaultCosmetics()
+        val batch = db.batch()
+        defaults.forEach { item ->
+          val data =
+              mapOf(
+                  "id" to item.id,
+                  "name" to item.name,
+                  "price" to item.price,
+                  "imageResName" to drawableToImageResName(item.imageRes))
+          batch.set(shopItemsCol().document(item.id), data)
```
Comment (Adessrrrr) [inline app/src/main/java/com/android/sample/ui/shop/repository/FirestoreShopRepository.kt:131]: This sonar warning does not need to be addressed, code is fine like this.

File: app/src/main/java/com/android/sample/ui/shop/ShopViewModel.kt [hunk 3]
```diff
@@ -2,54 +2,62 @@ package com.android.sample.ui.shop
...
 // The assistance of an AI tool (ChatGPT) was solicited in writing this file.
 
+/**
+ * ViewModel for the Shop screen. Manages cosmetic items and purchase logic with Firebase
+ * persistence.
+ */
 class ShopViewModel(
-    private val profileRepository: ProfileRepository = ProfileRepositoryProvider.repository
+    private val profileRepository: ProfileRepository = ProfileRepositoryProvider.repository,
+    private val shopRepository: ShopRepository = AppRepositories.shopRepository
```
Comment (Adessrrrr) [inline app/src/main/java/com/android/sample/ui/shop/ShopViewModel.kt:22]: It's nice to have a dedicated shop repository called inside shop view model.

File: app/src/test/java/com/android/sample/ui/shop/repository/FakeShopRepositoryTest.kt
Comment (Adessrrrr) [inline app/src/test/java/com/android/sample/ui/shop/repository/FakeShopRepositoryTest.kt:1]: It's good that you also tested the fake repo.

File: app/src/test/java/com/android/sample/ui/shop/repository/FirestoreShopRepositoryTest.kt
Comment (Adessrrrr) [inline app/src/test/java/com/android/sample/ui/shop/repository/FirestoreShopRepositoryTest.kt:1]: Tests are good, nothing to say about it.

File: firebase/firestore/firestore.rules
Comment (Adessrrrr) [inline firebase/firestore/firestore.rules:1]: The new rules concerning the shop handling in Firestore look good.

**Review by Adessrrrr — CHANGES_REQUESTED** [review]
Overall this is a very good PR, it's something that was needed on the project since until now the shop was just using fake repositories, and the items that a person bought were not really stored inside the inventory.

But now with the owned items and the shop linked to Firestore, you implemented persistent ownership after buying an item, which was mandatory for the application, good job.

It is also a big step in fixing the bug in the shop where we can buy an object multiple times, it will be simpler to resolve now.

For the coverage and code smells I have nothing to say, no code smells on Sonar and a line coverage of 98.4% which is huge, this is very nice. There are only some hardcoded strings and magic numbers in this PR that need to be changed, but otherwise it's perfect.

So, after changing the hardcoded strings and magic numbers, this PR can be merged, as it is very very good now.


(PS: there is duplicated code but since it's only 2.2 % we can accept it, because the rest of the PR is very good.)

File: app/src/main/java/com/android/sample/ui/shop/repository/FakeShopRepository.kt (hunk 1 above)
Comment (khlilinfo) [reply app/src/main/java/com/android/sample/ui/shop/repository/FakeShopRepository.kt]: this has been taken care of

**Review by Adessrrrr — APPROVED** [review]
I can see that all the requested changes were done, thanks for the work.

This PR can be merged now.
